### PR TITLE
Fixes sleep (STOP mode) on Photon platform

### LINKED ISSUE #655: [Photon]: System.sleep(_pin, _edge) never wakes
There is a similar issue for Core ( #640 ).
There is some confusion, since the docs state this should work, but it doesn't.
I've tested all pins with all FALLING/RISING/CHANGE and apart from some CHANGE combinations none of the other combinations did wake the Photon - not even WKP.

I tested with this code
```
void setup() 
{
  pinMode(D7, OUTPUT);
  Particle.function("GoToSleep", go2sleep);
}

void loop() 
{
    digitalWrite(D7, (millis() >> 2) & 0x88);
}

int go2sleep(String cmd)
{
    int i;
    if (i = cmd.toInt())
    {
        int pin = abs(i) % 20;  // 20 denotes pin D0
        PinMode input = INPUT;
        InterruptMode edge = CHANGE;

        if (abs(i) < 100) // >= 100 CHANGE
        {
            input = (i > 0) ? INPUT_PULLDOWN : INPUT_PULLUP;
            edge = (i > 0) ? RISING : FALLING;
        }
        pinMode(pin, input);
        System.sleep(pin, edge);
    }
    return i;
}

```
I also realized, that after the few CHANGE wakes that worked, the code continued from within the `Particle.function()` which does not fit the docs
```
When the specific interrupt arrives, the device awakens from stop mode, it will behave 
as if the device is reset and run all user code from the beginning with no values being 
maintained in memory from before the stop mode.
```
Although the note states
```
(Note: The new Particle Photon firmware will not reset before going into stop mode so 
all the application variables are preserved after waking up from this mode. The voltage 
regulator is put in low-power mode. This mode achieves the lowest power consumption 
while retaining the contents of SRAM and registers.)
```
What is meant with "new firmware"? A version number should be stated.
And also that the code carries on from where it left off (preserved registers doesn't mean too much to average users - as instruction and stack pointer would not do either)

### PR DESCRIPTION
**Pull request by avtolstoy**
HAL_Core_Enter_Stop_Mode should be wrapped with a critical section (cpsid/cpsie) to avoid
having an interrupt triggered while the device is entering sleep mode. Interrupts can be safely
disabled while in STOP mode.
All EXTI interrupts but the one associated with selected wakeup pin should be disabled as well
and restored after exiting STOP mode.

Fixes #655

### CODE REVIEW

**Comment by avtolstoy** [comment]
I have updated the PR. Waking up on RTC should now work correctly on Photon.
b92448f0c21055fb5219f9d1437dfd2cb6f1d2bc also correctly suspends networking before entering STOP mode and restores it after waking up.

**Comment by m-mcgowan** [comment]
Tested on Core, Photon, Electron with a test app:

 - registers a `sleep` function with the cloud, arg is the number of seconds to sleep.  The sleep call is `System.sleep(A1, RISING, seconds)`
 - loop() writes the current millis() count to serial every 100ms (no delays, polls millis())
 - A1 connected to GND

Results:

- calling `sleep 10`, the device shutdown the network, then slept for 10 seconds. After waking up, the network/cloud reconnected and serial output resumed. (On the Core, the serial connection terminates on sleep, since the device resets as part of executing stop mode.)
- calling `sleep 1000` the device sleeps for a long time. Wakes up when connecting A1 to 3v3.  No other gpio pins wake the device.
- after waking up, verified that the setup button continues to work
- tried this 10 times.

Outlier on the Electron, U270 rev15:

- when the A1 pin is floating, the electron would not sleep but wake up immediately. When this happened the setup button no longer continued to function. This was quite hard to reproduce reliably, and didn't happen when the pin was connected to gnd/3v3.

**Comment by m-mcgowan** [comment]
As you can see from the test above, this works well on all our platforms. Nice job @avtolstoy :+1:

File: hal/src/stm32f2xx/core_hal_stm32f2xx.c [hunk 1]
```diff
@@ -364,37 +367,84 @@ void HAL_Core_Enter_Safe_Mode(void* reserved)
     HAL_Core_System_Reset();
 }
 
-void HAL_Core_Enter_Stop_Mode(uint16_t wakeUpPin, uint16_t edgeTriggerMode)
+void HAL_Core_Enter_Stop_Mode(uint16_t wakeUpPin, uint16_t edgeTriggerMode, long seconds)
 {
+    if (!((wakeUpPin < TOTAL_PINS) && (wakeUpPin >= 0) && (edgeTriggerMode <= FALLING)) && seconds <= 0)
```
Comment (m-mcgowan) [inline hal/src/stm32f2xx/core_hal_stm32f2xx.c]: I think this should be `|| seconds < 0` - I will add that edit in develop.

**Pull request merged**